### PR TITLE
Vet clinic database: database performance audi

### PR DESCRIPTION
**Pull request by ZuhRa-HashiMi**
**The first screenshot should show how much time the query takes BEFORE the improvements are introduced**

first_screen

**The second screenshot should show how much time the query takes AFTER the improvements are introduced.**

second_screenshot

**With the help of my partner  @Bornittah **
For this project we had special preparation. As the goal of this project is to solve some performance issue, first we need to introduce those issues. In order to do that, I populated my database with a significant number of data.

### CODE REVIEW

**Review by mckent05 — COMMENTED** [review]
Hi @ZuhRa-HashiMi, 

# STATUS: INVALID REQUEST ❌ ❌ 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that does not have a commit from your partner on this milestone. Please note that since this project is a pair-programming project, your partner needs to have at least one commit in this PR, please fix this 🙏 

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: schema.sql [hunk 1]
```diff
@@ -59,3 +59,6 @@ CREATE TABLE visits (
     vet_id INTEGER REFERENCES vets(id) ON UPDATE CASCADE ON DELETE CASCADE,
     visit_date DATE NOT NULL
 );
+
+-- Add an email column to your owners table
+ALTER TABLE owners ADD COLUMN email VARCHAR(120);
```
Comment (MaganaAsiati) [inline schema.sql]: - Kindly make sure that you add in your schema.sql the way to decrease the execution time for each query.  I can suggest you add indexes to solve that. follow this [link](https://dataschool.com/sql-optimization/how-indexing-works/) to learn more bout indexing :+1: 

```
e.g  
-- Add animals_id index to visits table
CREATE INDEX visits_animals_id_asc ON visits(animal_id ASC);
-- delete visits_animals_id_asc  index 
 DROP INDEX  visits_animals_id_asc;
-- Create vets_id index on visits table
CREATE INDEX visits_vets_id_asc ON visits(vet_id ASC);
-- delete  visits_vets_id_asc index 
DROP INDEX  visits_vets_id_asc;
-- CREATE email index on owners table
CREATE INDEX owners_email_asc ON owners(email ASC);
-- delete  visits_vets_id_asc index 
DROP INDEX  owners_email_asc;
```

**Review by MaganaAsiati — CHANGES_REQUESTED** [review]
# STATUS: Required Changes ♻️♻️♻️ 

Hi @ZuhRa-HashiMi @mckent05 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

AlonzoLeroneThisCloseGIF

## Highlights
✅ No linter error
✅ README is professional

## Required Changes ♻️

- Depending on your machine's speed, it might be enough or not. Check that by running explain analyze SELECT COUNT(*) FROM visits where animal_id = 4: - If you get Execution time: X ms and X >= 1000: that should be enough, you can continue to the project requirements. - If you get Execution time: X ms and X < 1000: please go back to point 3. and repeat this query below until you get a value bigger than 1000ms. as you it show in the picture you take the execution time still less than 1000 please fix that

  ```sql

  -- This will add 3.594.280 visits considering you have 10 animals, 4 vets, and it will use around ~87.000 timestamps (~4min approx.)
  INSERT INTO visits (animal_id, vet_id, date_of_visit) SELECT * FROM (SELECT id FROM animals) animal_ids, (SELECT id FROM vets) vets_ids, generate_series('1980-01-01'::timestamp, '2021-01-01', '4 hours') visit_timestamp;

  -- This will add 2.500.000 owners with full_name = 'Owner <X>' and email = 'owner_<X>@email.com' (~2min approx.)
  insert into owners (full_name, email) select 'Owner ' || generate_series(1,2500000), 'owner_' || generate_series(1,2500000) || '@mail.com';
  ```

- [ ] Kindly included a pair of screenshots in the PR description with the results of the `EXPLAIN ANALYZE` command for each query (Screenshots should contain a total of 6 queries): :+1: 

* The first screenshot should show how much time the query takes BEFORE the improvements are introduced. - In the query planner output you should see the usage of `Seq scan` and/or `Filter` - The execution time should be around 1 sec (1000 ms). :+1: 
* The second screenshot should show how much time the query takes AFTER the improvements are introduced. - The execution time should be significantly lower than 1 sec (1000 ms). :+1: 

- [ ] It is a good practice to ensure you have a valid license properly set up on your projects. The MIT License link on your README file links to a 404 page. Kindly add a valid LICENSE. You can follow the instructions on the right side of the article attached below to set up a valid [LICENSE](https://choosealicense.com/licenses/mit/) :+1: 
- [ ] Kindly Update the readme file by adding the informations of your collaborator and removing the unused features to make your readme look clean and more readable :+1: 



_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

File: schema.sql (hunk 1 above)
Comment (t-yanick) [reply schema.sql]: > * Kindly make sure that you add in your schema.sql the way to decrease the execution time for each query.  I can suggest you add indexes to solve that. follow this [link](https://dataschool.com/sql-optimization/how-indexing-works/) to learn more bout indexing 👍
> 
> ```
> e.g  
> -- Add animals_id index to visits table
> CREATE INDEX visits_animals_id_asc ON visits(animal_id ASC);
> -- delete visits_animals_id_asc  index 
>  DROP INDEX  visits_animals_id_asc;
> -- Create vets_id index on visits table
> CREATE INDEX visits_vets_id_asc ON visits(vet_id ASC);
> -- delete  visits_vets_id_asc index 
> DROP INDEX  visits_vets_id_asc;
> -- CREATE email index on owners table
> CREATE INDEX owners_email_asc ON owners(email ASC);
> -- delete  visits_vets_id_asc index 
> DROP INDEX  owners_email_asc;
> ```

Above is a quote of the previous request.

File: schema.sql [hunk 2]
```diff
@@ -59,3 +59,6 @@ CREATE TABLE visits (
     vet_id INTEGER REFERENCES vets(id) ON UPDATE CASCADE ON DELETE CASCADE,
     visit_date DATE NOT NULL
 );
+
```
Comment (t-yanick) [inline schema.sql:62]: - [ ] As requested by previous reviewer, add indexes in schema.sql to reduce excution time for each query. Kindly do this if you don't mind.

**Review by t-yanick — CHANGES_REQUESTED** [review]
Hi @ZuhRa-HashiMi and @Bornittah , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## To Highlight
:heavy_check_mark:  Good job working on changes requested by previous reviewer
:heavy_check_mark: Commits from both team members. Good collaboration
:heavy_check_mark: Descriptive PR with requested performance screenshots.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by KayLemba — APPROVED** [review]
[# STATUS: APPROVED :green_circle: :clap: 

Hello  @ZuhRa-HashiMi , 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: aaa
Congratulations! 🎉 

HIGHLIGHTS:

- [x]  Your repo is professional 
- [x] Descriptive P.R








 








Now you can merge this branch and move on :1st_place_medal: 
Happy Coding cheers 🥇 💯 



well-1

Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._]

**Pull request merged**